encodings/dict/src/array.rs

```diff
@@ -97,8 +102,25 @@ impl CanonicalVTable<DictArray> for DictEncoding {
                 let canonical_values: Array = array.values().into_canonical()?.into_array();
                 take(canonical_values, array.codes())?.into_canonical()
             }
-            // Non-string case: take and then canonicalize
-            _ => take(array.values(), array.codes())?.into_canonical(),
+            DType::Primitive(ptype, _)
+                // TODO(alex): handle nullable codes & values
+                if array.values().len() <= 32 && *ptype != PType::F16
+                    && !array.codes().dtype().is_nullable()
+                    && !array.values().dtype().is_nullable() =>
+            {
+                let codes = array.codes().into_primitive()?;
+                let values = array.values().into_primitive()?;
+
+                match_each_unsigned_integer_ptype!(codes.ptype(), |$C| {
+                    match_each_native_simd_ptype!(values.ptype(), |$V| {
+                        // SIMD types larger than the SIMD register size are beneficial for
+                        // performance as this leads to better instruction level parallelism.
+                        let decoded = dict_decode_typed_primitive::<$C, $V, 32>(codes.as_slice(), values.as_slice());
+                        try_cast(decoded, array.dtype())?.into_canonical()
```

Comment:
you don't need the cast here anymore, values and codes have the same nullability